### PR TITLE
nfs: fail if no mds are running

### PR DESCRIPTION
**Pull request by leseb**
**Description of your changes:**

In order to run ganesha exporting cephfs we must have an MDS server
running. So, if not present we fail the instanciation of the CR.

Closes: https://github.com/rook/rook/issues/3835
Signed-off-by: Sébastien Han <seb@redhat.com>

**Which issue is resolved by this Pull Request:**
Resolves https://github.com/rook/rook/issues/3835

**Checklist:**

- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

### CODE REVIEW

**Comment by leseb** [comment]
> How about if we also query ceph for the existence of the pool specified in `n.Spec.RADOS.Pool`?

Done.

File: pkg/operator/ceph/nfs/nfs.go [hunk 1]
```diff
@@ -240,5 +241,23 @@ func validateGanesha(context *clusterd.Context, n cephv1.CephNFS) error {
...
+	}
+	if len(filesystems.Items) == 0 {
+		return fmt.Errorf("a cephfilesystem cr must be present to run nfs ressources")
+	}
+
+	// Second validation after the presence of an MDS server
+	// This time, we check the existence of the pool specified in n.Spec.RADOS.Pool
+	_, err = client.GetPoolDetails(context, n.Namespace, n.Spec.RADOS.Pool)
+	if err != nil {
+		return fmt.Errorf("pool %s not found, did the filesystem cr successfully completed? %+v", n.Spec.RADOS.Pool, err)
```
Comment (travisn) [inline pkg/operator/ceph/nfs/nfs.go]: nit: "successfully complete?"

**Comment by leseb** [comment]
> A separate PR for another day is to retry creating the NFS. If they create the filesystem.yaml and nfs.yaml at the same time, the nfs setup should eventually retry and succeed instead of always failing.

You mean, if a filesystem resource exists but no pool is found?

**Comment by travisn** [comment]
> You mean, if a filesystem resource exists but no pool is found?

Even if the filesystem CR doesn't exist yet, they might create it soon, in which case we could retry. We have a similar retry pattern for creating object users where we retry at least for a few minutes

File: pkg/operator/ceph/nfs/nfs.go [hunk 2]
```diff
@@ -240,5 +241,23 @@ func validateGanesha(context *clusterd.Context, n cephv1.CephNFS) error {
 		return fmt.Errorf("at least one active server required")
 	}
 
+	// We cannot run an NFS server if no MDS is running
+	// We cannot wait for the pool to exist in addRADOSConfigFile()
+	// but this will hold the orchestration and does not help much even a CephFilesystems CRD is injected
+	filesystems, err := context.RookClientset.CephV1().CephFilesystems(n.Namespace).List(metav1.ListOptions{})
```
Comment (BlaineEXE) [inline pkg/operator/ceph/nfs/nfs.go]: Doesn't this only list the filesystems the user wants? This doesn't guarantee that MDSes are running right?

Comment (travisn) [reply pkg/operator/ceph/nfs/nfs.go]: 1. The check for the pool existing below is really the one that will verify that the necessary pool exists. 
1. NFS can also be run on top of an object store as mentioned in [this line](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/nfs.yaml#L10), right? That means that checking for the pool to exist is the only check we can rely on.

**Pull request merged**